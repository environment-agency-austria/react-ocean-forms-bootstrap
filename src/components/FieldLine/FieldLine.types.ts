import { IUseFieldResult } from 'react-ocean-forms';

/**
 * Props for components that use the info
 * feature of the FieldLine
 */
export interface IInfoProps {
  /**
   * Info text
   */
  info?: string;
}

/**
 * Base props for the FieldLine component
 */
export interface IBaseFieldLineProps extends IInfoProps {
  /**
   * Label of the input element
   */
  label: string;
  /**
   * Label size for the bootstrap label column
   */
  labelSize?: string;
  /**
   * Input size for the bootstrap input column
   */
  inputSize?: string;
  /**
   * CssClass for the label column
   */
  labelClass?: string;
  /**
   * CssClass for the field line container
   */
  containerClass?: string;

  /**
   * Prefix component
   */
  prefix?: React.ReactNode;
  /**
   * Suffix component
   */
  suffix?: React.ReactNode;
}

/**
 * Props for the FieldLine component
 */
export interface IFieldLineProps extends IUseFieldResult<unknown>, IBaseFieldLineProps { }
