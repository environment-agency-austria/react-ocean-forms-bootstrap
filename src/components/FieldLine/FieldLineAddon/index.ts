export * from './FieldLineAddon';
export * from './FieldLineAddon.types';
