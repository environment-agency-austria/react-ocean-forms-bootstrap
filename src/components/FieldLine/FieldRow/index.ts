export * from './FieldRow';
export * from './FieldRow.types';
