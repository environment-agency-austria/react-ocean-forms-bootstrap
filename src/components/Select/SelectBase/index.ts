export * from './SelectBase';
export * from './SelectBase.types';
