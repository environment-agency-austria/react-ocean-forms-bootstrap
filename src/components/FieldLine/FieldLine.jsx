/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
