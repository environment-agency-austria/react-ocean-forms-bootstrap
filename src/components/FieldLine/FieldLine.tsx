/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText, PropsOf } from 'react-ocean-forms';
import { Col, InputGroup, Label } from 'reactstrap';

import { FieldError } from './FieldError';
import { FieldLineAddon } from './FieldLineAddon';
import { FieldRow } from './FieldRow';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import { InvalidAlert } from './InvalidAlert';
import { RequiredMarker } from './RequiredMarker';
import { ValidatingSpinner } from './ValidatingSpinner';

import { IFieldLineProps } from './FieldLine.types';

interface IFieldLineState {
  infoVisible: boolean;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export class FieldLine extends React.Component<IFieldLineProps, IFieldLineState> {
  public static displayName: string = 'FieldLine';

  public static defaultProps = {
    labelSize: '3',
    inputSize: '9',
    labelClass: 'text-right',
  };

  constructor(props: IFieldLineProps) {
    super(props);

    this.state = {
      infoVisible: false,
    };
  }

  /**
   * Toggles the visibility of the info alert
   */
  private toggleInfo = (): void => {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  public render(): JSX.Element {
    const {
      field,
      meta,
      label,
      children,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

export type DefaultizedFieldLineProps = JSX.LibraryManagedAttributes<typeof FieldLine, PropsOf<typeof FieldLine>>;
