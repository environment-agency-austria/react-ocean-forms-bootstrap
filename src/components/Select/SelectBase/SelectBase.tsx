/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { useCallback, useMemo } from 'react';
import { Input as StrapInput } from 'reactstrap';

import { useField } from 'react-ocean-forms';
import { ActionMeta } from 'react-select/lib/types';
import { FieldLine } from '../../FieldLine';
import {
  IPreparedSelectProps, ISelectBaseProps, ISelectFieldValue,
  ISelectOption, ISelectOptions, isSelectOption,
} from './SelectBase.types';

/**
 * The type that the field can have
 */
type SelectFieldValue = ISelectFieldValue | undefined;

/**
 * Result of an update
 */
interface IUpdatedLabelsResult {
  /**
   * True, if options were updated
   */
  needsUpdate: boolean;
  /**
   * The options
   */
  selectedOptions: SelectFieldValue;
}

/**
 * Base component for displaying bootstrap
 * form groups with an select input and
 * oForm support
 */
export const SelectBase = <TSubmitValue extends unknown = ISelectFieldValue>(props: ISelectBaseProps<TSubmitValue>): JSX.Element => {
  const {
    renderSelect,
    placeholder = 'ojs_select_placeholder',
    options,

    // booleans
    multi = false,
    loading,
    rtl,
    searchable,
    clearable,

    // Other props (field line)
    ...fieldlineProps
  } = props;

  const { fieldProps, metaProps } = useField(props);

  /**
   * Manually handle the onChange event of
   * the select because oForms is expecting
   * default input onChange behaviour.
   */
  const handleChange = useCallback((value: ISelectFieldValue, action?: ActionMeta): void => {
    const { handleChange } = props;

    fieldProps.onChange({
      target: {
        value,
      },
    });

    if (handleChange) {
      handleChange(value, action);
    }
  }, [fieldProps, props]);

  /**
   * Manually handle the onBlur event of
   * the select because oForms is expecting
   * default input onBlur behaviour.
   */
  const handleBlur = useCallback((): void => {
    fieldProps.onBlur();
  }, [fieldProps]);

  /**
   * Update the label of a select option according to the available options.
   * If null is returned, the label is already up to date.
   *
   * @param options The available select options
   * @param value The value that may need update
   */
  const updateSelectOptionLabel = useCallback((options: ISelectOptions, value: ISelectOption): ISelectOption | undefined => {
    let selectableValue: ISelectOption | undefined;
    options.some(item => {
      if (item.value !== value.value) { return false; }

      selectableValue = item;

      return true;
    });

    if (selectableValue === undefined || selectableValue.label === value.label) {
      return undefined;
    }

    return {
      ...value,
      label: selectableValue.label,
    };
  }, []);

  /**
   * Get the selected options with updated labels.
   * This might be needed if the labels were updated because of a language changes
   * @param options The available options
   * @param value The current value
   */
  const getUpdateLabels = useCallback((options: ISelectOptions, value: SelectFieldValue): IUpdatedLabelsResult => {
    let selectedOptions: SelectFieldValue;
    let needsUpdate = false;

    if (isSelectOption(value)) {
      // If only one option is selected
      const updated = updateSelectOptionLabel(options, value);
      needsUpdate = (updated !== undefined);
      selectedOptions = (needsUpdate)
        ? updated
        : value;
    } else if (Array.isArray(value)) {
      // Map the values
      selectedOptions = value.map(v => {
        const updated = updateSelectOptionLabel(options, v);
        // If null was returned, no updated is needed
        if (updated === undefined) { return v; }
        // Otherwise set the flag to true
        needsUpdate = true;

        return updated;
      });
    }

    return {
      needsUpdate,
      selectedOptions,
    };
  }, [updateSelectOptionLabel]);

  /**
   * Get the value of the field with updated labels
   * @param field The field
   * @param options The available options
   */
  const fieldValue = useMemo(() => {
    const fieldValue = (fieldProps.value as ISelectFieldValue | undefined);
    const {
      needsUpdate,
      selectedOptions,
    } = getUpdateLabels(options, fieldValue);
    if (needsUpdate && selectedOptions !== undefined) {
      // Check if the current value has a different label than the value
      // with the same key in the options array. Bugfix to change the
      // selected label when the current language changes.
      handleChange(selectedOptions);
    }

    return selectedOptions;
  }, [fieldProps.value, getUpdateLabels, handleChange, options]);

  // Support for plaintext display
  if (metaProps.plaintext) {
    let displayValue = '';
    if (isSelectOption(fieldValue)) {
      displayValue = fieldValue.label;
    } else if (Array.isArray(fieldValue)) {
      displayValue = fieldValue.map(v => v.label).join(', ');
    }

    return (
      <FieldLine {...props} fieldProps={fieldProps} metaProps={metaProps}>
        <StrapInput
          {...fieldProps}
          value={displayValue}
          plaintext
          onChange={undefined}
        />
      </FieldLine>
    );
  }

  // Generate a css class based on the validity of the select element
  let selectClass = 'react-select-control';
  selectClass = (selectClass + (metaProps.valid ? '' : ' is-invalid')).trim();

  // Format the placeholder
  const placeholderText = metaProps.stringFormatter(placeholder);

  // Prepare the props for the renderSelect method
  const preparedProps: IPreparedSelectProps = {
    id: fieldProps.id,
    inputId: `${fieldProps.id}-input`,
    value: fieldValue,
    isDisabled: fieldProps.disabled,
    isMulti: multi,
    isRtl: rtl,
    isLoading: loading,
    isSearchable: searchable,
    isClearable: clearable,
    onChange: handleChange,
    onBlur: handleBlur,
    options: options,
    placeholder: placeholderText,
    className: selectClass,
  };

  return (
    <FieldLine
      fieldProps={fieldProps}
      metaProps={metaProps}
      {...fieldlineProps}
    >
      {renderSelect(preparedProps, fieldProps)}
    </FieldLine>
  );
};
