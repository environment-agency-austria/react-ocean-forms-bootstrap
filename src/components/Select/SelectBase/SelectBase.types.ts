/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IUseFieldProps, IFieldComponentFieldProps } from 'react-ocean-forms';
import { Props as SelectProps } from 'react-select/lib/Select';
import { ActionMeta } from 'react-select/lib/types';

import { IBaseFieldLineProps } from '../../FieldLine';

/**
 * Selectable option
 */
export interface ISelectOption {
  /**
   * Display label
   */
  label: string;
  /**
   * Internal value
   */
  value: string;
}

export function isSelectOption(object: any): object is ISelectOption {
  return object && typeof (object as ISelectOption).label === 'string' && typeof (object as ISelectOption).value === 'string';
}

/**
 * Type for an array of ISelectOption
 * @see ISelectOption
 */
export type ISelectOptions = ISelectOption[];

/**
 * Types for the value that the field can have
 */
export type ISelectFieldValue = ISelectOption | ISelectOption[];

/**
 * Wrapper for the base props that maps the props from react-select
 */
interface IMappedReactSelectProps {
  /**
   * Selectable options
   */
  options: ISelectOptions;
  /**
   * Placeholder string that will be displayed while
   * the component is empty
   */
  placeholder?: string;
  /**
   * True, if the select component should be in multi-select mode
   */
  multi?: boolean;
  /**
   * Renders the component from right-to-left
   */
  rtl?: boolean;
  /**
   * If true, the component shows an loading state
   */
  loading?: boolean;
  /**
   * True if the component is searchable, false otherwise.
   */
  searchable?: boolean;
  /**
   * True if the component is clearable, false otherwise.
   */
  clearable?: boolean;
  /**
   * custom code for onChange
   */
  handleChange?(value: ISelectFieldValue, action?: ActionMeta): void;
}

export interface ISelectBasePropsBase<TSubmitValue> extends
  IUseFieldProps<ISelectFieldValue, TSubmitValue>,
  IBaseFieldLineProps,
  IMappedReactSelectProps {
}

/**
 * The props that are prepared by the `SelectBase` component
 * and passed to the `renderSelect` method
 */
export type IPreparedSelectProps = SelectProps;

/**
 * Props for the `SelectBase` component
 */
export interface ISelectBaseProps<TSubmitValue> extends ISelectBasePropsBase<TSubmitValue> {
  /**
   * Callback to render the actual select component with `preparedProps`
   * @param preparedProps The prepared props for the select
   */
  renderSelect(preparedProps: IPreparedSelectProps, fieldProps: IFieldComponentFieldProps<ISelectFieldValue>): JSX.Element;
  handleChange?(value: ISelectFieldValue, action?: ActionMeta): void;
}
