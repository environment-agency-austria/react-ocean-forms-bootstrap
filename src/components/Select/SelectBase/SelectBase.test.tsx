import * as React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { IFieldComponentFieldProps, IFieldComponentMeta } from 'react-ocean-forms';

import { createMockField, createMockFieldMeta } from '../../../test-utils/enzymeFormContext';
import { FieldLine } from '../../FieldLine';
import { ISelectOption, ISelectOptions } from '../SelectBase';
import { SelectBase } from './SelectBase';
import { IPreparedSelectProps, ISelectBaseProps } from './SelectBase.types';

describe('<SelectBase />', () => {
  interface ISetupArgs {
    props?: Partial<ISelectBaseProps>;
    fieldOverrides?: Partial<IFieldComponentFieldProps>;
    metaOverrides?: Partial<IFieldComponentMeta>;
  }

  interface ISetupResult {
    wrapper: ShallowWrapper;
    field: IFieldComponentFieldProps;
    meta: IFieldComponentMeta;
    options: ISelectOptions;
  }

  const setup = ({
    props,
    metaOverrides,
    fieldOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const fieldLabel = 'field0';
    const meta = {
      ...createMockFieldMeta(),
      ...metaOverrides,
    };
    const field = {
      ...createMockField(),
      value: '',
      ...fieldOverrides,
    };

    const options = [
      { value: 'one', label: 'One' },
      { value: 'two', label: 'Two' },
    ];

    const renderSelect = jest.fn((preparedProps: IPreparedSelectProps) => {
      return (
        <div>
          mock select with props
          {JSON.stringify(preparedProps)}
        </div>
      );
    });

    const wrapper = shallow((
      <SelectBase
        label={fieldLabel}
        meta={meta}
        field={field}
        options={options}
        renderSelect={renderSelect}
        {...props}
      />
    ));

    return {
      wrapper,
      field,
      meta,
      options,
    };
  };

  it('should render correctly', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a field line and a react-select', () => {
    const renderSelect = jest.fn();

    const { wrapper } = setup({
      props: {
        renderSelect,
      },
    });

    expect(wrapper.find(FieldLine).exists()).toBeTruthy();
    // expect(wrapper.find(ReactSelect).exists()).toBeTruthy();
    expect(renderSelect).toBeCalled();
  });

  it('should pass the meta props to the field line', () => {
    const { wrapper, meta } = setup();

    expect(wrapper.find(FieldLine).prop('meta')).toEqual(meta);
  });

  describe('label edge case', () => {
    it('should update the label of the selected value if it does not match with the label of the item inside options', () => {
      const { field } = setup({
        fieldOverrides: { value: { value: 'two', label: '' } },
      });

      const fieldValue: ISelectOption = field.value as ISelectOption;
      expect(fieldValue.label).toBe('');
    });

    it('should update the label of the selected values if it does not match with the label of the item inside options', () => {
      const { field } = setup({
        props: { multi: true },
        fieldOverrides: { value: [{ value: 'two', label: '' }] },
      });

      const fieldValue: ISelectOptions = field.value as ISelectOptions;
      expect(fieldValue[0].label).toBe('');
    });

    it('should trigger the field.onChange event if the labels mismatch', () => {
      const { field } = setup({
        fieldOverrides: { value: { value: 'two', label: '' } },
      });

      expect(field.onChange).toHaveBeenCalledWith({
        target: {
          value: { value: 'two', label: 'Two' },
        },
      });
    });
  });

  describe('field events', () => {
    it('should call field.onChange when the input changes', () => {
      let changeProp: Function | undefined;
      const renderSelect = jest.fn().mockImplementation((props: IPreparedSelectProps) => {
        changeProp = props.onChange;
      });

      const { field, options } = setup({
        props: { renderSelect },
      });

      expect(changeProp).toBeDefined();
      changeProp && changeProp(options[1]);

      expect(field.onChange).toHaveBeenCalledWith({
        target: {
          value: options[1],
        },
      });
    });

    it('should call field.onChange when the input changes (multi=true)', () => {
      let changeProp: Function | undefined;
      const renderSelect = jest.fn().mockImplementation((props: IPreparedSelectProps) => {
        changeProp = props.onChange;
      });

      const { field, options } = setup({
        props: {
          renderSelect,
          multi: true,
        },
      });

      const selected = [options[0], options[1]];

      expect(changeProp).toBeDefined();
      changeProp && changeProp(selected);

      expect(field.onChange).toHaveBeenCalledWith({
        target: {
          value: selected,
        },
      });
    });

    it('should call field.onBlur when there is an input blur', () => {
      let blurProp: Function | undefined;
      const renderSelect = jest.fn().mockImplementation((props: IPreparedSelectProps) => {
        blurProp = props.onBlur;
      });

      const { field } = setup({
        props: { renderSelect },
      });

      expect(blurProp).toBeDefined();
      blurProp && blurProp();

      expect(field.onBlur).toHaveBeenCalledWith();
    });
  });

  describe('meta.plaintext', () => {
    it('should correctly react to meta.plaintext', () => {
      const { wrapper } = setup({
        metaOverrides: { plaintext: true },
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('should display the label property of the selected option in plaintext mode', () => {
      const { wrapper } = setup({
        metaOverrides: { plaintext: true },
        fieldOverrides: { value: { value: 'two', label: 'Two' } },
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('should display the label property of the selected option in plaintext mode (multi=true)', () => {
      const { wrapper } = setup({
        props: { multi: true },
        metaOverrides: { plaintext: true },
        fieldOverrides: { value: [
          { value: 'one', label: 'One' },
          { value: 'two', label: 'Two' },
        ] },
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should set the is-invalid css class to the ReactSelect component if meta.valid is false', () => {
    let className: string | undefined;
    const renderSelect = jest.fn().mockImplementation((props: IPreparedSelectProps) => {
      className = props.className;
    });

    setup({
      metaOverrides: { valid: false },
      props: {
        renderSelect,
      },
    });

    expect(className).toBe('react-select-control is-invalid');
  });

  describe('handleChange', () => {
    it('render with handleChange and call handleChange method', () => {
      let changeProp: Function | undefined;
      const renderSelect = jest.fn().mockImplementation((props: IPreparedSelectProps) => {
        changeProp = props.onChange;
      });
      const handleChange = jest.fn();

      const { options } = setup({
        props: {
          renderSelect,
          handleChange,
        },
      });

      const selected = [options[0], options[1]];
      expect(changeProp).toBeDefined();
      changeProp && changeProp(selected);

      expect(handleChange).toBeDefined();
      expect(handleChange).toBeCalled();
    });
  });
});
