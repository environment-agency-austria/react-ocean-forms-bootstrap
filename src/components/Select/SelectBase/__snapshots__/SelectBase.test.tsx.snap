// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectBase /> meta.plaintext should correctly react to meta.plaintext 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "mock-field",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": "",
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": true,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  multi={false}
  options={
    Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
    ]
  }
  placeholder="ojs_select_placeholder"
  renderSelect={[MockFunction]}
>
  <Input
    disabled={false}
    id="mock-field"
    name="mock-field"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    plaintext={true}
    type="text"
    value=""
  />
</FieldLine>
`;

exports[`<SelectBase /> meta.plaintext should display the label property of the selected option in plaintext mode (multi=true) 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "mock-field",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": Array [
        Object {
          "label": "One",
          "value": "one",
        },
        Object {
          "label": "Two",
          "value": "two",
        },
      ],
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": true,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  multi={true}
  options={
    Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
    ]
  }
  placeholder="ojs_select_placeholder"
  renderSelect={[MockFunction]}
>
  <Input
    disabled={false}
    id="mock-field"
    name="mock-field"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    plaintext={true}
    type="text"
    value="One, Two"
  />
</FieldLine>
`;

exports[`<SelectBase /> meta.plaintext should display the label property of the selected option in plaintext mode 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "mock-field",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": Object {
        "label": "Two",
        "value": "two",
      },
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": true,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
  multi={false}
  options={
    Array [
      Object {
        "label": "One",
        "value": "one",
      },
      Object {
        "label": "Two",
        "value": "two",
      },
    ]
  }
  placeholder="ojs_select_placeholder"
  renderSelect={[MockFunction]}
>
  <Input
    disabled={false}
    id="mock-field"
    name="mock-field"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
    plaintext={true}
    type="text"
    value="Two"
  />
</FieldLine>
`;

exports[`<SelectBase /> should render correctly 1`] = `
<FieldLine
  field={
    Object {
      "disabled": false,
      "id": "mock-field",
      "name": "mock-field",
      "onBlur": [MockFunction],
      "onChange": [MockFunction],
      "value": "",
    }
  }
  inputSize="9"
  label="field0"
  labelClass="text-right"
  labelSize="3"
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": false,
      "stringFormatter": [MockFunction] {
        "calls": Array [
          Array [
            "ojs_select_placeholder",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "ojs_select_placeholder",
          },
        ],
      },
      "touched": false,
      "valid": true,
    }
  }
>
  <div>
    mock select with props
    {"id":"mock-field","inputId":"mock-field-input","isDisabled":false,"isMulti":false,"options":[{"value":"one","label":"One"},{"value":"two","label":"Two"}],"placeholder":"ojs_select_placeholder","className":"react-select-control"}
  </div>
</FieldLine>
`;
